### PR TITLE
Make Consensus.CoinbaseMaturity property immutable

### PR DESCRIPTION
**Pull request by rowandh**

### CODE REVIEW

File: src/Stratis.Bitcoin.IntegrationTests/TestRegTest.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿using Stratis.Bitcoin.Networks;
+using Stratis.Bitcoin.Tests.Common;
+
+namespace Stratis.Bitcoin.IntegrationTests
+{
+    /// <summary>
+    /// This network enables an immutable Consensus object for testing only.
```
Comment (zeptin) [inline src/Stratis.Bitcoin.IntegrationTests/TestRegTest.cs]: Mutable

File: src/Stratis.Bitcoin.IntegrationTests/TestRegTest.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    /// </summary>
+    internal class TestRegTest : BitcoinRegTest
```
Comment (zeptin) [inline src/Stratis.Bitcoin.IntegrationTests/TestRegTest.cs]: Will we need a TestStratisRegTest as well?

File: src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletAddressGenerationAndFundsVisibility_Steps.cs [hunk 3]
```diff
@@ -34,14 +35,14 @@ public partial class Wallet_address_generation_and_funds_visibility : BddSpecifi
...
-            this.nodeGroupBuilder = new NodeGroupBuilder(Path.Combine(this.GetType().Name, this.CurrentTest.DisplayName), KnownNetworks.RegTest);
+            this.nodeGroupBuilder = new NodeGroupBuilder(Path.Combine(this.GetType().Name, this.CurrentTest.DisplayName), NetworkRegistration.Register(new TestRegTest()));
             this.sharedSteps = new SharedSteps();
         }
 
         protected override void AfterTest()
         {
-            this.sendingStratisBitcoinNode.FullNode.Network.Consensus.CoinbaseMaturity = this.previousCoinBaseMaturity;
-            this.receivingStratisBitcoinNode.FullNode.Network.Consensus.CoinbaseMaturity = this.previousCoinBaseMaturity;
+            ((MutableTestConsensus) this.sendingStratisBitcoinNode.FullNode.Network.Consensus).CoinbaseMaturity = this.previousCoinBaseMaturity;
```
Comment (zeptin) [inline src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletAddressGenerationAndFundsVisibility_Steps.cs]: Are these the only integration tests that currently fail if you use an immutable CoinBaseMaturity? I feel like there should be others.

**Comment by ferdeen** [comment]
Is the title to this - _PR Make Consensus.CoinbaseMaturity property **mutable**_ ?

File: src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletAddressGenerationAndFundsVisibility_Steps.cs (hunk 3 above)
Comment (rowandh) [reply src/Stratis.Bitcoin.IntegrationTests/Wallet/WalletAddressGenerationAndFundsVisibility_Steps.cs]: These are the only ones that modify CoinBaseMaturity.

File: src/Stratis.Bitcoin.IntegrationTests/TestRegTest.cs (hunk 2 above)
Comment (rowandh) [reply src/Stratis.Bitcoin.IntegrationTests/TestRegTest.cs]: Possibly. There are other fields on Consensus that still have setters used in other tests.

File: src/Stratis.Bitcoin.IntegrationTests/TestRegTest.cs (hunk 1 above)
Comment (rowandh) [reply src/Stratis.Bitcoin.IntegrationTests/TestRegTest.cs]: Oops, nice catch!

File: src/Stratis.Bitcoin.Tests.Common/MutableTestConsensus.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        public int CoinType { get; set; }
+        public BigInteger ProofOfStakeLimit { get; set; }
+        public BigInteger ProofOfStakeLimitV2 { get; set; }
+        public int LastPOWBlock { get; set; }
+        public bool IsProofOfStake { get; set; }
+        public uint256 DefaultAssumeValid { get; set; }
+        public ConsensusFactory ConsensusFactory { get; set; }
+        public ICollection<IConsensusRule> Rules { get; set; }
+    }
+}
```
Comment (monsieurleberre) [inline src/Stratis.Bitcoin.Tests.Common/MutableTestConsensus.cs]: Please can you order those fields alphabetically ?
Kidding obviously 😂

**Pull request closed without merging**